app/src/main/kotlin/com/hippo/ehviewer/ui/main/RollingNumber.kt

```diff
@@ -0,0 +1,154 @@
+import androidx.compose.ui.layout.layout
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.text.TextStyle
+import androidx.compose.ui.text.rememberTextMeasurer
+import androidx.compose.ui.unit.IntOffset
+import androidx.compose.ui.unit.IntSize
+import androidx.compose.ui.unit.TextUnit
+import androidx.compose.ui.unit.dp
+import kotlin.math.PI
+import kotlin.math.atan2
+import kotlin.math.cos
+import kotlin.math.floor
+import kotlin.math.max
+import kotlin.math.sin
+
+@Composable
+fun RollingNumber(number: Int, style: TextStyle = LocalTextStyle.current, width: Int? = null) {
+    val string = remember(number) { "$number" }
+    val transition = updateTransition(string)
+    val textMeasurer = rememberTextMeasurer()
+    val size by transition.animateIntSize { str ->
+        remember(str) { textMeasurer.measure(str, style).size }
+    }
+    val density = LocalDensity.current
+    val (styleNoSpacing, spacing) = remember(density, style) {
+        if (style.letterSpacing != TextUnit.Unspecified) {
+            val spacing = with(density) { style.letterSpacing.toDp() }
+            style.copy(letterSpacing = TextUnit.Unspecified) to spacing
+        } else {
+            style to 0.dp
+        }
+    }
+    Row(
+        modifier = Modifier.clipToBounds().layout { measurable, constraints ->
+            val placeable = measurable.measure(constraints)
+            layout(width = placeable.width, height = size.height) {
+                placeable.place(0, 0)
+            }
+        },
+        horizontalArrangement = Arrangement.End,
+    ) {
+        val content = remember {
+            val meta = @Composable { reversed: Int ->
+                val max = width ?: with(transition) { max(currentState.length, targetState.length) }
+                val rotate by transition.animateOffset { str ->
+                    val len = str.length
+                    val absent = max - len
+                    val where = max - reversed - 1
+                    val v = if (where < absent) null else str[where - absent].digitToInt()
+                    mapSimplyConnectedElementToEuclideanPartialCircle(v)
+                }
+                Column(
+                    modifier = Modifier.offset {
+                        val number = partialCircleToLinearUIOffset(rotate)
+                        numberToOffset(size, 9 - number)
+                    },
+                    horizontalAlignment = Alignment.CenterHorizontally,
+                ) {
+                    repeat(10) { i ->
+                        Text(
+                            text = "${9 - i}",
+                            style = styleNoSpacing,
+                        )
+                        Text(
+                            text = "",
+                            style = styleNoSpacing,
+                        )
+                    }
+                }
+            }
+            (0 until maxNumber).map {
+                movableContentOf { meta(it) }
```

Comment:
No need to use `movableContent` if width is fixed.

Reply:
Done